shell/platform/tizen/channels/text_input_channel.cc

```diff
@@ -310,48 +311,58 @@ void TextInputChannel::HandleUnfilteredEvent(Ecore_Event_Key* event) {
     return;
   }
 #endif
-
   bool select = !strcmp(event->key, "Select");
-  bool is_filtered = true;
-  if (!strcmp(event->key, "Left")) {
-    if (active_model_->MoveCursorBack()) {
-      SendStateUpdate(*active_model_);
-    }
-  } else if (!strcmp(event->key, "Right")) {
-    if (active_model_->MoveCursorForward()) {
-      SendStateUpdate(*active_model_);
-    }
-  } else if (!strcmp(event->key, "End")) {
-    if (active_model_) {
-      active_model_->MoveCursorToEnd();
-      SendStateUpdate(*active_model_);
+  bool shift = event->modifiers & ECORE_SHIFT;
+  bool need_update = false;
+  std::string key = event->key;
+
+  if (key == "Left") {
+    if (shift) {
+      TextRange selection = active_model_->selection();
+      need_update = active_model_->SetSelection(
+          TextRange(selection.base(), selection.extent() - 1));
+    } else {
+      need_update = active_model_->MoveCursorBack();
     }
-  } else if (!strcmp(event->key, "Home")) {
-    if (active_model_) {
-      active_model_->MoveCursorToBeginning();
-      SendStateUpdate(*active_model_);
+  } else if (key == "Right") {
+    if (shift) {
+      TextRange selection = active_model_->selection();
+      need_update = active_model_->SetSelection(
+          TextRange(selection.base(), selection.extent() + 1));
+    } else {
+      need_update = active_model_->MoveCursorForward();
     }
-  } else if (!strcmp(event->key, "BackSpace")) {
-    if (active_model_->Backspace()) {
-      SendStateUpdate(*active_model_);
+  } else if (key == "End") {
+    if (shift) {
+      need_update = active_model_->SelectToEnd();
+    } else {
+      need_update = active_model_->MoveCursorToEnd();
     }
-  } else if (!strcmp(event->key, "Delete")) {
-    if (active_model_->Delete()) {
-      SendStateUpdate(*active_model_);
+  } else if (key == "Home") {
+    if (shift) {
+      need_update = active_model_->SelectToBeginning();
+    } else {
+      need_update = active_model_->MoveCursorToBeginning();
     }
-  } else if (!strcmp(event->key, "Return") ||
-             (select && !text_editing_context_.is_in_select_mode_)) {
-    EnterPressed(active_model_.get(), select);
-
+  } else if (key == "BackSpace") {
+    need_update = active_model_->Backspace();
+  } else if (key == "Delete") {
+    need_update = active_model_->Delete();
   } else if (event->string && strlen(event->string) == 1 &&
              IsASCIIPrintableKey(event->string[0])) {
     active_model_->AddCodePoint(event->string[0]);
-    SendStateUpdate(*active_model_);
+    need_update = true;
+  } else if (!strcmp(event->key, "Return") ||
+             (select && !text_editing_context_.is_in_select_mode_)) {
+    EnterPressed(active_model_.get(), select);
+    return;
   } else {
-    is_filtered = false;
+    FT_LOG(Warn) << "Key[" << key << "] is unhandled!";
```

Comment:
Please don't use exclamation marks when it's not really important.

Reply:
Ok, I removed it